### ISSUE TITLE
Nested dependent role run with the same variable

### ISSUE THREAD

**Issue by clement-escolano** [ansible/ansible#74733]
### Summary

When two roles depends on a nested dependent role (with different variables passed), the nested role is only run once with the latest value specified.

### Issue Type

Bug Report

### Component Name

roles

### Ansible Version

```console
$ ansible --version
ansible 2.10.9
  config file = /redacted/ansible.cfg
  configured module search path = ['/Users/someuser/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/3.4.0/libexec/lib/python3.9/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.9.5 (default, May  4 2021, 03:29:30) [Clang 11.0.0 (clang-1100.0.33.17)]
```

### Configuration

```console
$ ansible-config dump --only-changed
DEFAULT_ROLES_PATH(/redacted/ansible.cfg) = ['/redacted/roles']
DEFAULT_VAULT_PASSWORD_FILE(/redacted/ansible.cfg) = /redacted/.vault-password
INTERPRETER_PYTHON(/redacted/ansible.cfg) = auto
RETRY_FILES_ENABLED(/redacted/ansible.cfg) = True
```

### OS / Environment

OS is not relevant.

### Steps to Reproduce

Here is the example playbook:
```yaml
# roles/app1/meta/main.yml
---
dependencies:
  - role: my-dependency
    vars:
      myvar: app1

```

```yaml
# roles/app2/meta/main.yml
---
dependencies:
  - role: my-dependency
    vars:
      myvar: app2

```

```yaml
# roles/my-dependency/meta/main.yml
---
dependencies:
  - role: nested-dependency
    vars:
      nestedvar: "{{ myvar }}"

```

```yaml
# roles/nested-dependency/tasks/main.yml
---
- debug: msg="nested-dependency {{ nestedvar }}"

```

```yaml
# roles/nested-dependency/meta/main.yml
---
allow_duplicates: yes

```

```yaml
# play.yml
---
- name: Basic configuration
  hosts: all
  roles:
    - app1
    - app2
```

Run:

```
ansible-playbook play.yml -i hosts.yml
```

### Expected Results

I expect `nested-dependency` role to be run twice, once with `nestedvar` being `app1` and once with `nestedvar` being `app2`.

### Actual Results

```console
`nested-dependency` role is run twice but with `app2` both times.
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by sivel** [comment]
Duplicate of:

* https://github.com/ansible/ansible/issues/72650
* https://github.com/ansible/ansible/issues/67718

**Comment by clement-escolano** [comment]
Hello @sivel.
Thank you for taking the time to read me. This is indeed a perfect duplicate of #72650 sorry for not finding it out first (I did search the issues 🤷).
I think implementing #67718 may fix the issue but I think you pointed the issue that is happening here in #72650: `nestedvar: "{{ myvar }}"` is lazily evaluated so is considered identical by the role deduplication algorithm (hence running only once when `allow_duplicates` is `false`, and running with the same value when `allow_duplicates` is `true`). Do you know if eager evaluation of variable is something considered in ansible? It would fix this issue more properly I think.
Have a good day

**Comment by sivel** [comment]
> Do you know if eager evaluation of variable is something considered in ansible?

This is on our radar, but is not something that is being actively worked on.

**Comment by clement-escolano** [comment]
OK I see thank you.
About #67718, do you have any timeline for implementation (1 month, 1 year ?) or it is the same as eager evaluation?

**Comment by samdoran** [comment]
> do you have any timeline for implementation (1 month, 1 year ?)

Unknown at this point.

**Comment by clement-escolano** [comment]
Thank you for the information.
Have a good day
